avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGenerator.java

```diff
@@ -530,12 +552,13 @@ private void processUnion(JVar unionSchemaVar, final String name, final Schema u
         unionAction = FieldAction.fromValues(optionSchema.getType(), false, EMPTY_SYMBOL);
       }
 
-      ifBlock = ifBlock != null ? ifBlock._elseif(condition) : body._if(condition);
-      final JBlock thenBlock = ifBlock._then();
-
       JVar optionSchemaVar = null;
       if (useGenericTypes && unionAction.getShouldRead()) {
-        JInvocation optionSchemaExpression = unionSchemaVar.invoke("getTypes").invoke("get").arg(JExpr.lit(i));
+        if (-1 == readerOptionUnionBranchIndex) {
+          // TODO: Improve the flow of this code... it's messy needing to check for this here...
+          throw new FastSerdeGeneratorException("Unexpected internal state. The readerOptionUnionBranchIndex should be defined if unionAction.getShouldRead() == true.");
```

Comment:
When this "unexpected internal state" will happen? If so, the following `processComplexType` may also fail due to `Null` `processComplexType`?

Reply:
According to the current logic, the `unionAction.getShouldRead()` condition should be false if `action.getShouldRead()` was also false. If `action.getShouldRead()` is false, then `readerOptionUnionBranchIndex` would still be at -1, but that is fine since `unionAction.getShouldRead()` would prevent us from reaching here...

Furthermore, if we fail to set `readerOptionUnionBranchIndex` then we would also leave `readerOptionSchema == null`, which should likewise make use short-circuit the code here. So I think the current logic is correct, although as I mentioned in the TODO, the flow is quite convoluted and could lead to fragility when future maintenance happens...

I would like to refactor the flow to make things simpler and more robust, but given the time pressure to fix the union bugs, I'd like to minimize the scope of the change in the short-term.